MaterialLib/MPL/Property.h

```diff
@@ -0,0 +1,116 @@
+/// boost::variant container. It can hold scalars, vectors, tensors, and so
+/// forth.
+enum PropertyDataTypeName
+{
+    nScalar,
+    nPair,
+    nVector,
+    nSymmTensor,
+    nTensor
+};
+
+using PropertyDataType =
+    boost::variant<double, Pair, Vector, SymmTensor, Tensor, std::string>;
+
+/// This class is the base class for any material property of any
+/// scale (i.e. components, phases, media, ...). The single value of
+/// that Property can hold scalars, vectors, tensors, strings, etc.
+class Property
+{
+public:
+    virtual ~Property() = default;
+    /// This method is called when a property is used for the wrong kind of
+    /// material, or if the property is not implemented on this kind of material
+    /// yet.
+    void notImplemented(const std::string& property,
+                        const std::string& material) const;
+    /// This virtual method simply returns the private _value attribute without
+    /// changing it.
+    virtual PropertyDataType value() const;
+    /// This virtual method will compute the property value based on the primary
+    /// variables that are passed as arguments.
+    virtual PropertyDataType value(VariableArray const& variable_array) const;
+    /// This virtual method will compute the derivative of a property
+    /// with respect to the given variables pv.
+    virtual PropertyDataType dValue(VariableArray const& variable_array,
+                                    Variables const variables) const;
+    /// This virtual method will compute the second derivative of a
+    /// property with respect to the given variables pv1 and pv2.
+    virtual PropertyDataType d2Value(VariableArray const& variable_array,
+                                     Variables const variables1,
+                                     Variables const variables2) const;
+
+    template <typename T>
+    T value() const
+    {
+        return boost::get<T>(value());
+    }
+    template <typename T>
+    T value(VariableArray const& variable_array) const
+    {
+        return boost::get<T>(value(variable_array));
+    }
+    template <typename T>
+    T dValue(VariableArray const& variable_array,
+             Variables const variables) const
+    {
+        return boost::get<T>(dValue(variable_array, variables));
+    }
+    template <typename T>
+    T d2Value(VariableArray const& variable_array,
+              Variables const& variables1,
+              Variables const& variables2) const
+    {
+        return boost::get<T>(d2Value(variable_array, variables1, variables2));
+    }
+
+protected:
+    /// The single value of a property.
+    PropertyDataType _value;
+    PropertyDataType _dvalue;
+};
+
+/// This method returns the 0-based index of the variant data types. Can be
+/// enhanced by using enums.
+inline std::size_t getType(Property const& p)
+{
+    return p.value().which();
+}
+
+inline void overwriteExistingProperties(PropertyArray& properties,
```

Comment:
Since in line 111 `new_properties` entries are moved, the `new_properties` aren't valid anymore, isn't it? Maybe the name of the function is a bit misleading, but I haven't a good suggestion.

Reply:
Good point. Need a name.